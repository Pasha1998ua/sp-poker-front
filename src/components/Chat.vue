<template>
  <div class="chat-block">
    <button @click="emitMessage">Send</button>
    <input v-model="inputVal">
    <table class="chat-list">
      <tr v-for="(chatMsg,index) in messagePool" :key="index">
        <td>{{ chatMsg.sentTimestamp }}</td>
        <td>{{ chatMsg.userNickName }}</td>
        <td>{{ chatMsg.message }}</td>
      </tr>
    </table>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { mapActions, mapGetters } from 'vuex'
import ChatMessage from '@/models/ChatMessage'
import localStorageHelper from '@/helpers/localStorageHelper'

export default defineComponent({
  name: 'MainPage',
  data () {
    return {
      inputVal: '' as string
    }
  },
  computed: {
    ...mapGetters({
      messagePool: 'getMessagePool'
    })
  },
  created () {
    this.listenMessages()
  },
  beforeUnmount () {
    this.stopListening()
  },
  methods: {
    ...mapActions({
      listenMessages: 'listenMessages',
      stopListening: 'stopListening',
      sendMessage: 'sendMessage'
    }),
    emitMessage () {
      const chatMsg: ChatMessage = {
        room: 'PavloKey',
        userNickName: localStorageHelper.getUserName(),
        sentTimestamp: Date.now(),
        message: this.inputVal
      }
      this.sendMessage(chatMsg)
    }
  }
})
</script>

<style scoped lang="scss">
.chat-list {
  max-width: 600px;
  margin: 20px;
  border-collapse: collapse;

  td, th {
    padding: 0.5rem;
    border: 1px solid black;
  }
}
</style>
