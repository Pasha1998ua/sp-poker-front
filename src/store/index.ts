import { createStore } from 'vuex'
import SocketHelper from '@/helpers/socketHelper'
import ChatMessage from '@/models/ChatMessage'

const SOCKET_HELPER = SocketHelper.getInstance()

export default createStore({
  state: {
    messagePool: [] as Array<ChatMessage>
  },
  mutations: {
    addMessage (state, chatMsg: ChatMessage) {
      state.messagePool.push(chatMsg)
    }
  },
  actions: {
    listenMessages ({ commit }) {
      SOCKET_HELPER.listenMessages((chatMsg: ChatMessage) => {
        commit('addMessage', chatMsg)
      })
    },
    stopListening () {
      SOCKET_HELPER.stopListening()
    },
    // eslint-disable-next-line
    sendMessage ({ commit }, chatMsg: ChatMessage) {
      SOCKET_HELPER.sendMessage(chatMsg)
    }
  },
  getters: {
    getMessagePool (state): Array<ChatMessage> {
      return state.messagePool
    }
  },
  strict: true
})
