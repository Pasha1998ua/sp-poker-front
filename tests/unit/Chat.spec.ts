import { shallowMount } from '@vue/test-utils'
import Chat from '@/components/Chat.vue'
import store from '@/store'

describe('Chat.vue', () => {
  it('renders page correctly', () => {
    const wrapper = shallowMount(Chat, {
      global: {
        mocks: {
          $store: store
        }
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
