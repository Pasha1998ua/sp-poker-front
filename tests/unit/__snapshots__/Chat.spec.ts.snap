// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat.vue renders page correctly 1`] = `
<div class="chat-block"><button>Send</button><input>
  <table class="chat-list"></table>
</div>
`;
